dlfcn.go

```diff
@@ -17,20 +18,66 @@ const RTLD_GLOBAL = 0x8
 
 const RTLD_DEFAULT = ^uintptr(1)
 
-func Dlopen(name string, mode int) uintptr {
-	bs := strings.CString(name)
+// Dlopen examines the dynamic library or bundle file specified by path. If the file is compatible
+// with the current process and has not already been loaded into the
+// current process, it is loaded and linked. After being linked, if it contains
+// any initializer functions, they are called, before Dlopen
+// returns. It returns a handle that can be used with Dlsym and Dlclose.
+// A second call to Dlopen with the same path will return the same handle, but the internal
+// reference count for the handle will be incremented. Therefore, all
+// Dlopen calls should be balanced with a Dlclose call.
+func Dlopen(path string, mode int) uintptr {
+	bs := strings.CString(path)
 	ret, _, _ := SyscallN(dlopenABI0, uintptr(unsafe.Pointer(bs)), uintptr(mode), 0)
 	runtime.KeepAlive(bs)
 	return ret
 }
 
+// Dlsym takes a "handle" of a dynamic library returned by Dlopen and the symbol name.
+// It returns the address where that symbol is loaded into memory. If the symbol is not found,
+// in the specified library or any of the libraries that were automatically loaded by Dlopen
+// when that library was loaded, Dlsym returns zero.
 func Dlsym(handle uintptr, name string) uintptr {
 	bs := strings.CString(name)
 	ret, _, _ := SyscallN(dlsymABI0, handle, uintptr(unsafe.Pointer(bs)), 0)
 	runtime.KeepAlive(bs)
 	return ret
 }
 
+// Dlerror returns a human-readable string describing the most recent error that
+// occurred from Dlopen, Dlsym or Dlclose since the last call to Dlerror. It
+// returns an empty string if no errors have occurred since initialization or
+// since it was last called.
+func Dlerror() string {
+	msg, _, _ := SyscallN(dlerrorABI0)
+	if msg == 0 {
+		return ""
+	}
+	var length int
+	for {
+		// use unsafe.Add once we reach 1.17
+		//_ = (*byte)(unsafe.Add(unsafe.Pointer(msg), length))
+		if *(*byte)(unsafe.Pointer(msg + uintptr(length))) == '\x00' {
+			break
+		}
+		length++
+	}
+	// use unsafe.Slice once we reach 1.17
+	// we make a slice then cast to a string to force Go to copy the string from C memory into Go memory
+	return string(*(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{Data: msg, Len: length, Cap: length})))
```

Comment:
What is the lifetime of msg?

Reply:
It is valid until the next call to dlerror. Which is why we copy it into Go so that Go can decide when it no longer needs the string

Reply:
Does this line really copy the string? Probably at casting from []byte to string?

Reply:
Yeah a cast []byte to string copies bc it can’t prove that the []byte won’t be modified and break the immutability of string

Reply:
I went ahead and made the copying more obvious